p4-16/psa/PSA.mdk

```diff
@@ -2212,46 +2212,59 @@ program could discard the most significant bits of the timestamp so
-digest message. The receiver identifier is an opaque id used by the
-control plane to route the digest message to the right receiver. The
-digest message can contain any from the data-plane.
-
-In PSA, digest is performed at the end of the ingress pipeline in the
-deparser.
-~ TBD
-The digest extern provides a `pack` method to specify the
-content in the digest header.
-~
-
-The compiler decides the best serialization format to send the data
-object to the control plane.
 
-On the control plane side, the control plane API calls the `unpack`
-method, and pass in a reference to the data object to be read.
-
-The `unpack` method return 0 if succeed. The field names in the
-control plane API is derived from the field names in the data-plane
-`pack` method.
-
-### Control Plane
-
-The control plane should provide a
+## Packet Digest
 
+A digest is one mechanism to send a message from the data plane to the
+control plane.  Another is to send a packet to the control plane via
+the port numbered `PSA_PORT_CPU`.  Sending a packet to port
+`PSA_PORT_CPU` typically sends most or all of the original packet
+headers, and perhaps also the payload, each as a separate message to
+be received and processed by the control plane.  The contents of a
+digest for one packet are typically much smaller than the packet.  A
+PSA implementation can take advantage of this, e.g. it might combine
+digests for multiple packets into larger messages, to reduce the
+rate of messages sent to the control plane.
+
+A digest contains two pieces of information: a name, which is derived
+from the name of the digest object instance in the P4 program, and the
+digest message.  The digest message can contain any values from the
+data plane.
+
+In PSA, a digest is created by calling the `pack` method on the digest
+instance.  The argument is the value to be included in the digest,
+often a collection of values in a P4 `struct` type.  The compiler
+decides the best serialization format to send the digest contents to a
+local software agent, which is responsible for sending the digest data
+in a form to be defined by the P4 Runtime API working group.
+
+A PSA program can instantiate multiple Digest instances in the same
+ingress control block, and make at most one `pack` call on each
+instance during a single execution of the ingress control block.  The
+same applies independently for the egress control block.  The control
+plane can distinguish each of these messages by the name of the digest
+instance in the P4 program.
+
+Since high speed PSA implementations are expected to be able to
+generate digests much faster than control software can consume them,
+it is expected that loss of such digest messages will occur if the
+data plane generates them too quickly.  It is recommended that PSA
+implementations maintain a count of digest messages that the data
+plane requests, but do not reach the control plane, independently for
+each digest instance.
```

Comment:
Does this imply that the control plane needs to acknowledge the request of a digest? Or did you mean that the PSA data plane keeps a counter of digests, and the control plane subtracts the ones it received? Is this a counter per instance? How many instances do we allow? Are these counters just like table counters or maintained by the hardware in the data plane?

Reply:
Note: Providing these drop counters is a recommendation in the proposed text, not a requirement.  That said, I wouldn't mention them if nobody would ever find them useful, e.g. I expect they could be very useful for someone debugging a system, including control plane software, who is wondering why the system is misbehaving in some way, and trying to narrow down the cause of the problem.

I don't think such counters implies that the control plane needs to ack digest requests at all.

There are multiple ways to implement such loss counters, depending on the implementation, but for example, if you have an ASIC and local driver/agent software involved in getting digest messages to a controller, implementing a drop count in the ASIC, and a separate drop count in the driver/agent software, and adding up those two numbers, should do the job.

The text proposes "independently for each digest instance".

The number of Digest instances supported by an implementation I expect would be implementation-dependent, just as the number of tables, number of Counter, Meter, and Register instances, number of classes of service, number of ports, etc. is implementation-dependent.

These drop counters I would expect to be one of the cheapest parts of a Digest extern implementation.  Far more expensive I expect would be on-ASIC buffers to store multiple digest messages, and/or DMA hardware to get them to the local driver software at a reasonably high rate, without dropping too many when they might occur in short bursts.

Reply:
Thanks for the details. I'm still getting used to the fact that you consider the driver/agent part of the "hardware" and it is fair game to get it involved.

Reply:
I think of the driver/agent as a part of the system where digest notification messages might have to be dropped, and so if you want to know how many were dropped in the system between the source and the controller, it is best to count that reason, too.  A different kind of PSA implementation might only have one place where they could get dropped.